UniformCylindricalEndcap: more tests and improvement.

markscheel opened the pull request:
## Proposed changes
Fixes #4172 

The only lines of untested code should now be inside `if`s for things that should never happen unless there is a bug, and that I don't even know how to make happen (e.g. if the root is not bracketed [after I call code that should bracket the root], or if the root finder throws an exception, or if the map is not invertible [after I have checked all the cases so that the map should indeed be invertible]).

Also replaced some duplicated code with a function, and refactored a test to be slightly smaller.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

markscheel committed:
Refactor TestMapHelpers::test_suite_for_map_on_cylinder

markscheel committed:
UniformCylindricalEndcap: add roundoff test.

This tests the map for cases when the point is within
roundoff of the +/- z boundary of the cylinder.  These
are edge cases that trigger pieces of the code that
otherwise are untouched by the tests.

markscheel committed:
UniformCylindricalEndcap: remove duplicate code.

Creates a function that is called twice instead of duplicating
code.  One problem with the duplicated code is that one of
the duplicates had an `if` that was never tested (because
that `if` could not actually occur inside that function).

markscheel committed:
UniformCylindricalEndcap: add more inverse-map tests.

These make sure codecov sees the cases where the
inverse map automatically returns std::optional{} because
certain sanity checks are not satisfied.

markscheel commented:
The clang-tidy errors appear to be `toms748`-related.

wthrowe commented:
I ran the test a bunch of times and got the failure
```
/home/wthrowe/spectre/tests/Unit/Domain/CoordinateMaps/Test_UniformCylindricalEndcap.cpp:237: FAILED:
  {Unknown expression after the reported line}
due to unexpected exception with messages:
  UniformCylindricalEndcap
  Seed is: 3478862725 from /home/wthrowe/spectre/tests/Unit/Domain/
  CoordinateMaps/Test_UniformCylindricalEndcap.cpp:60
  center_two := { -0.55049704087110252, 0.38721696485721946, -0.
  74352714220913829 }
  radius_two := 7.00125814302648575
  z_plane_two := 5.32555645603222505
  radius_one := 0.67567060496540776
  z_plane_one := 4.46090043557018845
  center_one := { -1.83783820332234504, -0.54079760207445438, 4.
  18968968360496596 }
  Seed is: 1732363388 from /home/wthrowe/spectre/tests/Unit/Helpers/Domain/
  CoordinateMaps/TestMapHelpers.hpp:552
  height := -0.67365694762912809
  phi := 4.29601520369562806
  radius := 0.00235196860030478
  roundoff := 0.00000000000000088
  Test inverse map
  test_point := { -0.00095124597416826, -0.00215102008206532, 1.
  00000000000000089 }

[...]

  Line: 625 of /home/wthrowe/spectre/src/Domain/CoordinateMaps/
  UniformCylindricalEndcap.cpp
  Function: std::optional<std::array<double, 3>> domain::CoordinateMaps::
  UniformCylindricalEndcap::inverse(const std::array<double, 3> &) const
  Root is not bracketed: 0 7.37774e-05 0.00235197 1.02509e-14
  ############ ERROR ############
```

markscheel committed:
UniformCylindricalEndcap: check more bracket-fail cases.

In the inverse function, if the root isn't bracketed, there is
a check that the bracketing failure is caused by roundoff.
This check was done only if the function value was close to
zero.

This commit changes what 'close' means, and does the check much
more often.  Also this commit puts the value of 'close' into
a constexpr variable instead of hardcoding it into two 'if'
statements.

markscheel commented:
Added a new commit that fixes the roundoff-bracketing-failure problem.

wthrowe commented:
Still getting failures.
```
/home/wthrowe/spectre/tests/Unit/Domain/CoordinateMaps/Test_UniformCylindricalEndcap.cpp:237: FAILED:
  {Unknown expression after the reported line}
due to unexpected exception with messages:
  UniformCylindricalEndcap
  Seed is: 55135103 from /home/wthrowe/spectre/tests/Unit/Domain/
  CoordinateMaps/Test_UniformCylindricalEndcap.cpp:60
  center_two := { -0.97415220790710533, -0.04268721292324351, -0.
  56284487393790772 }
  radius_two := 10.28890043334250137
  z_plane_two := 6.77671036317570152
  radius_one := 5.73774986432597078
  z_plane_one := 5.458745287159112
  center_one := { -3.15290739311211254, 1.39832693426118082, 1.
  24847746012156513 }
  Seed is: 1194289024 from /home/wthrowe/spectre/tests/Unit/Helpers/Domain/
  CoordinateMaps/TestMapHelpers.hpp:552
  height := -0.65200235547932284
  phi := 2.33714402768342033
  radius := 0.00017242191494504
  roundoff := 0.0000000000000007
  Test inverse map
  test_point := { -0.00011957607687484, 0.00012422108755174, 1.
  00000000000000067 }
  
############ ERROR ############
  Shortened stack trace is:
[...]
  End shortened stack trace.
  
Node: 0 Proc: 0
  Line: 633 of /home/wthrowe/spectre/src/Domain/CoordinateMaps/
  UniformCylindricalEndcap.cpp
  Function: std::optional<std::array<double, 3>> domain::CoordinateMaps::
  UniformCylindricalEndcap::inverse(const std::array<double, 3> &) const
  Root is not bracketed: 0 1.89753e-06 0.000172422 2.3273e-14
  ############ ERROR ############
```

markscheel committed:
UniformCylindricalEndcap: keep z-planes farther apart.

Factor changes from 0.02 to 0.03.

markscheel commented:
@wthrowe I fixed your problem (by changing tolerance of bracked-fail case again) and ran the test 1000 times, and found and fixed another problem (fixed by not allowing geometry to be too extreme, in particular the two z-planes were allowed to be really close and I slightly increased the separation).  Reran the test 1000 times and it worked for me.

wthrowe reviewed (approved):
OK, this looks good now.

clang-tidy looks like all Boost stuff.  The macOS build failed for spack reasons unrelated to this PR, but I decided to rerun it just to be safe.
